### PR TITLE
Handle textDocument/foldingRange call

### PR DESCRIPTION
**Pull request by andfoy**
This PR enables pyls to handle and compute folding ranges for a given document

### CODE REVIEW

**Comment by andfoy** [comment]
@ccordoba12, this one is ready for review

File: pyls/hookspecs.py [hunk 1]
```diff
@@ -110,3 +110,8 @@ def pyls_settings(config):
 @hookspec(firstresult=True)
 def pyls_signature_help(config, workspace, document, position):
     pass
+
+
+@hookspec(firstresult=True)
+def pyls_folding_range(config, workspace, document):
```
Comment (gatesn) [inline pyls/hookspecs.py]: Can you move this alphabetically?

File: setup.py [hunk 2]
```diff
@@ -92,6 +92,7 @@
             'rope_completion = pyls.plugins.rope_completion',
             'rope_rename = pyls.plugins.rope_rename',
             'yapf = pyls.plugins.yapf_format',
+            'folding = pyls.plugins.folding'
```
Comment (gatesn) [inline setup.py]: Again, alphabetical

File: pyls/plugins/folding.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                    ctx, stack, line, col, folding_starts, folding_ends,
+                    tree_nodes, folding_ranges)
+
+    folding_ranges = sorted(folding_ranges)
+    ranges = __reduce_folding_ranges(folding_ranges)
+    return ranges
+
+
+@hookimpl
+def pyls_folding_range(document):
```
Comment (gatesn) [inline pyls/plugins/folding.py:16]: As a style point, can you move this above the "helper" functions.

File: pyls/plugins/folding.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    program = str(document.source)
+    # Add an additional "pass" to ensure that there's always a next node
+    program = program + '\n\npass'
```
Comment (gatesn) [inline pyls/plugins/folding.py]: Hah, clever

File: pyls/plugins/folding.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            offending_line = e.lineno
+            lines = program.splitlines(True)
+            lines = lines[:offending_line - 1]
+            program = ''.join(lines)
+            reparsing = True
+
+    # Parse again to add additional node
+    if reparsing:
+        program = program + '\n\npass'
+        tree = ast.parse(program)
```
Comment (gatesn) [inline pyls/plugins/folding.py]: Should we recurse instead, to handle the case when there are multiple syntax errors?

File: test/plugins/test_folding.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+                {'startLine': 39, 'endLine': 40},
+                {'startLine': 42, 'endLine': 43},
+                {'startLine': 45, 'endLine': 51},
+                {'startLine': 46, 'endLine': 47},
+                {'startLine': 48, 'endLine': 49},
+                {'startLine': 50, 'endLine': 51}]
+    assert ranges == expected
+
+
+def test_folding_syntax_error():
```
Comment (gatesn) [inline test/plugins/test_folding.py:148]: Let's test multiple syntax errors?

**Comment by andfoy** [comment]
@gatesn Do you agree in disabling code folding for syntax errors? We don't know how to handle this case with @ccordoba12

File: pyls/plugins/folding.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+from pyls import hookimpl
+
+SKIP_NODES = (tree_nodes.Module, tree_nodes.IfStmt, tree_nodes.TryStmt)
+IDENTATION_REGEX = re.compile(r'(\s+).+')
+
+
+@hookimpl
+def pyls_folding_range(document):
+    program = str(document.source) + '\n'
```
Comment (ccordoba12) [inline pyls/plugins/folding.py]: Is `str` really necessary here? I think it'd be better to write

    program = document.source + u'\n'
	
instead.

File: pyls/plugins/folding.py [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+    last_operator = isinstance(last_leaf, tree_nodes.Operator)
+    node_is_operator = isinstance(node, tree_nodes.Operator)
+    last_operator = last_operator or not node_is_operator
+
+    end_line -= 1
+
+    modified = False
+    if isinstance(node.parent, tree_nodes.PythonNode):
+        kind = node.type
+        # print(f'Parent node: {kind}')
```
Comment (ccordoba12) [inline pyls/plugins/folding.py]: Please remove this comment.

File: pyls/plugins/folding.py [hunk 9]
```diff
@@ -0,0 +1,169 @@
+# pylint: disable=len-as-condition
+
+import re
+
+import parso
+import parso.python.tree as tree_nodes
+
+from pyls import hookimpl
+
+SKIP_NODES = (tree_nodes.Module, tree_nodes.IfStmt, tree_nodes.TryStmt)
```
Comment (ccordoba12) [inline pyls/plugins/folding.py:11]: I think we should also provide folding on `if` and `try/except` statements. Why did you decide not to do that?

File: test/plugins/test_folding.py [hunk 10]
```diff
@@ -0,0 +1,163 @@
+
```
Comment (ccordoba12) [inline test/plugins/test_folding.py:2]: Palantir's copyright is missing here.

File: pyls/plugins/folding.py [hunk 11]
```diff
@@ -0,0 +1,169 @@
+# pylint: disable=len-as-condition
```
Comment (ccordoba12) [inline pyls/plugins/folding.py:1]: Please add Palantir's copyright above this line too.

**Review by ccordoba12 — COMMENTED** [review]
@andfoy, I left some small comments for you, otherwise looks good.

File: pyls/plugins/folding.py (hunk 9 above)
Comment (andfoy) [reply pyls/plugins/folding.py:11]: VSCode does individual if/else folding

Comment (ccordoba12) [reply pyls/plugins/folding.py:11]: I didn't know that. Then let's follow VSCode lead in this case.

**Pull request merged**